src/include/storage/index/index.h

```diff
@@ -50,46 +69,84 @@ class Index {
   virtual ~Index() = default;
 
   /**
-   * Inserts a new key-value pair into the index.
+   * Inserts a new key-value pair into the index, used for non-unique key indexes.
+   * @param txn txn context for the calling txn, used to register abort actions
    * @param tuple key
    * @param location value
    * @return false if the value already exists, true otherwise
    */
-  virtual bool Insert(const ProjectedRow &tuple, TupleSlot location) = 0;
+  virtual bool Insert(transaction::TransactionContext *txn, const ProjectedRow &tuple, TupleSlot location) = 0;
```

Comment:
From our experience with this method, there is a usability issue. For example, when we are building the index, we scan through the whole table, getting a bunch of `TupleSlot`s and related `ProjectedRow`s.

However, that projected row in the table can not be directly used in this `Insert` method (as well as others) since the internal metadata representation of "index projected rows" and "selected projected row from table" are different. I think the major reason is that indexes and tables are maintaining their own `col_oid`, `col_id`, actual layout id mappings.

Reply:
I think it could be nice to unify them or provide some handy functions.

Reply:
It's the execution layer's job to reason about passing well-formed buffers to the storage layer. Indexes don't know anything about tables or how to transform from a table's schema to the index's schema. In the case of indexes on expressions, you'd be pushing even more interpretation logic into the index wrapper. I don't think that's the right call. Let that be compiled code in the execution layer.

Reply:
Sounds reasonable.